### ISSUE TITLE
Improve container options to allow startup URLs to be combined with Erwic mode

### ISSUE THREAD

**Issue by helmling** [Jelmerro/Vieb#71]
What I want to do is something like: "vieb --erwic <json config for https://domain.com/a> https://domain.com/a/b"
and this should use the running erwic instance to open provided URL.

This would be the same behaviour as for the "normal" instance and according to the --help should already work.
The motivation is that I've configured my macOS to handle specific URL patterns with specific applications and would like to use vieb erwic instances for those applications now.

Is this possible?

Thanks for your help and thanks for this really cool browser.

**Comment by Jelmerro** [comment]
This is currently not possible, it will just read local JSON files. To achieve this, you can make a bash script to download the JSON file (such as with `wget` or `curl`) and then open it with `--erwic ./download.json` (given that you named the JSON file `download.json` in the current folder). I'll update the title of this issue to reflect a feature request, as this would be a nice improvement of the Erwic option.

**Comment by helmling** [comment]
Sorry, probably my question was not clear. What I what to do is the following:

1) create an applicaion named _myapp_ for domain "myapp.com" using erwic and e.g. shelll wrapper scirpt.
2) open this application _myapp_
3) configure my OS to us to open URL matching "https://myapp.com/*" using the applicaiion _myapp_ (This I did)
4) if the OS now open an URL like "https://myapp.com/path" the running appliaction "_myapp_" should display this URL.

That would be exactly the sam behaviour as when you configure your OS to open URL with vieb in general and according to what the --help option shows should already work (providing an URL together with the --erwic optoion)

Thanks a lot.

**Comment by Jelmerro** [comment]
Right, thanks for clearing that up, I misunderstood your first comment. Even though the `--help` docs aren't explicit about this, the `Erwic.md` docs are referring to the configured pages as a fixed set, and also mention that you can't add or remove these while running. One of the side effects is that startup URLs also won't be opened, as there is no container associated with them, and thus Erwic won't allow these URLs to open.

I'm not too sure why exactly you would want to use Erwic mode for your use case, as I don't see the difference with using Vieb regularly with `containertabs` enabled. I'll update the issue title again to reflect your actual request, but before working on this I first want to make sure what exactly would be it's use case.

**Comment by helmling** [comment]
The use case is the following: Asuming you have an erwic instance for gitbub.com which is running and displaying a page on github. You now might receive a URL pointing to github.com via mail, a message or any other channel.  It would be cool if clicking on this link would use the current running erwic instance of displaying it opening the default browser.
I acutally have a tool (hammerspoon) registerd as "default browser"  which dispatches URLs accding to patterns to different "browsers", so I can dispatch the link pointing to a page at github.com to the vieb erwic instance / application generated for githbug.
But this does not help unless there's a way to tell the github erwic instance to open the page.

So the goal is to to have different default browsers (== erwic instances) for differnt domains (-patterns)
Does this make sense to you? I currently use this scenario activly for a web based ticket system where I often receive links via mail.

**Comment by Jelmerro** [comment]
That makes sense, but I'm wondering which container the startup URLs should open. This is actually one of the questions I didn't have an answer to when I introduced Erwic, which resulted the blocking of new tabs being opened in Erwic mode altogether. Maybe we could consider using regular `containertabs` for both of these scenarios? They would show in yellow to indicate a temporary container, as opposed to containers that were opened by the JSON file. This won't actually solve your issue, as the newly opened tab won't share data with the already opened one, but it also seems weird to open a new tab and share data with the previous one, as that kind of defeats the purpose of Erwic mode.

**Comment by helmling** [comment]
Ah, now I see the problem. The reason why I completely missed this problem is that I don't have a use case for opening multiple URLS in one Erwic instance. I was rather seeing Erwic as a lightweight replacement for solutions like _epichrome_ or _natifyer_ which allow to have a web application behaving like a standalone application.

So coming back to your question: As long as one Erwic instance can open multiple websites the additional container tab might be the only solution I guess. (In case of limiting Erwic to one site/tab only it could be just a normal tab)

**Comment by helmling** [comment]
After thinking about it a bit longer I think the best would be if the content of the "matching" Erwic tab (app) would simply get replaced by the new URL. "matching" could mean one of the following (all implemntations would make sense):

- URL has same host/domain as the Erwic app
- URL starts with with the URL of the Erwic app
- the Erwic app has an additional property which is used to match the "starts with" of the provided URL

**Comment by Jelmerro** [comment]
We could also introduce a new setting in the Erwic JSON file to specify where it should open new tabs and also have options to use the same container as the currently opened tab. Maybe second setting similar to this can be added for what to do with startup pages. As an extension of this ticket I would like to improve the usability of containers, also allow users to create them whenever and give them custom names and colors. Below are just a couple of settings we could think of adding.

### containernewtab 

Set the name of the container to use when opening new tabs (Vieb default set to: `main`). Could also be set to any other container name, where `%n` will be replaced with an incrementing number and the `container` prefix will make the container be cleared on shutdown (similar to the current `containertabs` toggle, which would then become `containernewtab=container%n`). Although maybe this would be a nice moment to rename the `container` prefix to something else. Finally a special name could be added to open tabs externally, which would be set by default in Erwic mode: `containernewtab=external`, but can of course be changed with the usual `:set` command or the viebrc (inside the datafolder configured in the Erwic JSON file). This setting will affect all places where new tabs are opened, including follow mode for new tabs (which with this setting set to external, would open them in the default external browser instead, which is currently only the case in Erwic, but would now be a setting that can be optionally disabled in Erwic and optionally enabled in Vieb).

### containerstartuppage

Very similar to the previous setting, but applies to URLs that are opened as a startup argument. Unlike the `containernewtab` setting, the default is the same for both Vieb and Erwic, and would be set to `main`. This setting does not affect containers for which the name is already set in the Erwic JSON file.

### containercolors

I want this to be a list-like setting that maps container names (using regex) to colors. An example for the current behavior would be: `containercolors=main~#eee,container\d+~#ff0` although setting the main color would be a bit redundant, as the defaults for that won't be changed (while still allowing custom colors).

### Special container names

Below is a bulleted list of the things that make this setting different from just setting a basic name. These names or patterns offer additional functionality if this name or part of a name is used:

- `main` (default) not different from any other named container actually, but it's the default value of the settings
- `external` open the URLs in an external default browser (or other applications that opens URLs, which could be your "hammerspoon" program)
- `container...` if the name starts with "container" it will only be stored for the current session (again, this name might change in development)
- `%n` anywhere in the container name will be replaced with an automatically incrementing number (that starts 0 each session, though now that I think about this, it could also be replaced with some kind of random hash instead). This `%n` is most useful when combined with the `container` prefix above.

Hopefully you (and others interested) were able to read this long and detailed comment without too many issues. Please let me know if these settings would offer you with all the options you need for your described workflow. I'll definitely start work on some of these features already, but they can all be refined and changed. Ideally, these changes should appear in the 3.0.0 release, hopefully with just minor or no required changes.

**Comment by helmling** [comment]
I see this thing is getting complex :) and I struggled a little bit understanding your suggestion.
We have options (and a name) for the Eric instance and options (and names) for the apps inside this Erwic configured in the json file. Containernewtab and containerstartuppage would be we Eric instance options, right? 
I'm still confused. Is an Eric instance name the same as a container name? If yes the the Erwic instance name (in the json) should get renamed to conainername.

**Comment by Jelmerro** [comment]
No, these would be settings that can be used both in Vieb and in Erwic, just like any other setting. To summarize, they would allow you to specify the name of the container to use for the next tab. Sorry if that wasn't clear from my previous comment.

**Comment by helmling** [comment]
So maybe I misunderstood the Eric config. Is one Erwic instance with all the "apps" specified in the json running in one container or is there a container for each app, meaning the app-name would be the container name? I assumed the former.

**Comment by Jelmerro** [comment]
Erwic is the name of a single instance window that can open multiple apps/applications in a separate container each, based on the app name (=container name). The datafolder you specify is the only location Erwic will store data, so multiple instances are not connected in any way. Even within a single instance, all the apps you specify run in their own container (also known as session or partition). This will be unchanged after the suggestions above.

The suggestions in my previous comments are settings to make it explicit in which container newly opened tabs or startup pages should open.

The purpose of the `containerstartuppage` setting is to set the name of the container for opened startup pages (i.e. using the CLI or as an application argument). If you don't update this setting and keep the default, all startup pages will open in the `main` container. Pages opened using the Erwic config will still open in the container name specified (using the app name) and are not affected by this setting, because that would defeat the purpose of the config file. Additional URLs you specify WILL use this setting. For example, if using Erwic you only open a single app, it would make sense to set this setting to the same name as the app, so all pages opened from the URL arguments open in the same container. Otherwise the URL arguments will open in `main` by default.

The purpose of the `containernewtab` setting is to set the name of the container to use for tabs that are opened. In Vieb, this will also be set to `main`, which is basically the same as what's currently happening, but it's more explicit with this setting. In Erwic, this setting is by default set to `external`, which will redirect all URLs of newly opened tabs to the external default browser. This is also inline with the current Erwic behavoir. Because this is now a setting, you can enable the same behavior in Vieb, but also change this back to other options in Erwic. This last part is what you are interested in, as it will allow you to open new tabs in Erwic, if this setting is changed. For example, if you only have a discord app open, it would make sense to do this: `containernewtab=discord`, as all new tabs you open from within Erwic, will now open in a new tab of the discord container.

**Comment by Jelmerro** [comment]
To make it a bit more explicit what is actually being configured, the Erwic config is going through some changes. One of them is that the app "name" is replaced with a "container" field, as that's all the name was used for. The datafolder location will be moved to the startup arguments, so that it can also be used for regular Vieb startups (giving more flexibility compared to "--portable"), and so that the same Erwic config can be used for multiple instances in separate datafolders. You can find the new V3 version of the Erwic readme [here](https://github.com/Jelmerro/Vieb/blob/feature/v3/Erwic.md).

Besides that, most of the settings I outlined above are currently implemented on the V3 branch. The confusing "container" prefix has been replaced with "temp", short for temporary, as that outlines better what actually makes it different from regular containers. There are no docs for these new container settings on the V3 branch yet, but I'll work on those when everything is stabilized a bit more.

Lastly, I am thinking of adding more special container names to tailor your use case. This means names such as "replacematching" to find a matching tab to replace with the URL. It will take some time to think of which exact names to use for this type of functionality, but I'll keep you updated on the progress. Given the amount of special container names, I might start prefixing them to avoid confusion, but that's also something I still have to think about.

**Comment by helmling** [comment]
Thanks a lot for all this. I think it makes perfect sense to replace "app" with "container" in the json as it _is_ the same.
As far as In can see the new features should fullfill all my wishes - I don't want to run multiple containers in one Erwic and for that reason I dont need any config options for matching/finding containers to use for new tabs.

I'll checkout, build, play around with v3 and give you feedback.

Again, may thanks for considering and implementing my "feature request"

**Comment by helmling** [comment]
While testing I run into two issues (should I open new issues in the future or should I append to this?)
- build on macos fails while signing if the "extraFiles" are configured in the electron-builder.yml
  could get fixed by  changing extraFiles -> extraResources (or remove compleatly)
  (further the target should be dmg instead of zip for mac I guess)
- opening a new URL via the commandline (either in an erwic instance or in the "main" instance) crashes Vieb before starting
   with the new URL (I'll attache the macos crash report)
[dump.txt](https://github.com/Jelmerro/Vieb/files/5379330/dump.txt)
Beside that I could at least open new tabs in either the erwic window or the main window depending on the presents of the --erwic option in the arguments.

**Comment by Jelmerro** [comment]
I checked the build issue as I hadn't made a build yet while on the V3 branch. I don't actually get any warning at all from electron-builder. I suspect it's some apple stuff with regards to the signing again, which [I wouldn't be able to check, as I don't own a mac](https://github.com/electron-userland/electron-builder/issues/811). In one of my next commits I'll move the extraFiles section to the 'win' and 'linux' block specifically, so that hopefully mac will stop complaining about that.

As for DMG, [not going to happen in the near future](https://superuser.com/questions/226439/hdiutil-on-linux). The zip target should contain a mac .app, which is also a format your mac should be able to run. I have just checked again if there is a way to build other mac targets from ANY my machines, but there isn't, as apple is making it more and more difficult to release software for mac without owning one, which they plan to make required in the future.

Regarding the segfault crash, I'll try to reproduce it on my end. I haven't had this issue yet, but I haven't tested all the features on V3 in depth yet. If you can reproduce it, please make a new issue for it with steps to reproduce. I'll try to make it happen with the instructions you have listed in your comment.

**Comment by Jelmerro** [comment]
Improved containers and the settings explained above are implemented on the V3 branch.
They will be part of the 3.0.0 release that I plan to release soon after [Electron 11 goes stable (17 November)](https://www.electronjs.org/docs/tutorial/electron-timelines).

I wasn't able to reproduce the crash you mentioned, so please verify that it also happens after installing the latest Electron beta on the V3 branch. If so, please make a new issue with steps to reproduce it.
